### PR TITLE
[FLINK-9384]KafkaAvroTableSource failed to work due to type mismatch

### PR DESCRIPTION
**Pull request by tragicjun**
## What is the purpose of the change
An exception was thrown when using KafkaAvroTableSource, complaining like:
```
 "TableSource of type org.apache.flink.streaming.connectors.kafka.Kafka011AvroTableSource returned a DataStream of type GenericType<org.apache.flink.types.Row> that does not match with the type Row(id: Integer, name: String, age: Integer, event: GenericType<java.util.Map>) declared by the TableSource.getReturnType() method. Please validate the implementation of the TableSource."
```
This pull request fixes the issue by referencing the implementation of KafkaJsonTableSource. 

## Brief change log

  - Add a TypeInformation as the second argument to the constructor of AvroRowDeserializationSchema
  - AvroRowDeserializationSchema overrides the method getProducedType()
  - KafkaAvroTableSource constructs a TypeInformation converted from AvroRecordClass and pass it to AvroRowDeserializationSchema

## Verifying this change

This change is already covered by existing tests, such as KafkaAvroTableSourceTestBase.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (yes)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaAvroTableSource.java [hunk 1]
```diff
@@ -81,7 +85,12 @@ public String explainSource() {
 
 	@Override
 	protected AvroRowDeserializationSchema getDeserializationSchema() {
-		return new AvroRowDeserializationSchema(avroRecordClass);
+		return new AvroRowDeserializationSchema(avroRecordClass, tableSchemaToReturnType(schema));
+	}
+
+	/** Converts the table schema into into the return type. */
+	private static RowTypeInfo tableSchemaToReturnType(TableSchema tableSchema) {
```
Comment (suez1224) [inline flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaAvroTableSource.java]: Do you need this extra function, I think just inlining it should be fine.

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowDeserializationSchema.java [hunk 2]
```diff
@@ -110,6 +116,8 @@ private void writeObject(ObjectOutputStream oos) throws IOException {
 		oos.writeObject(recordClazz);
 	}
 
+
```
Comment (suez1224) [inline flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowDeserializationSchema.java]: extra blank line here.

**Comment by tragicjun** [comment]
Thanks @suez1224 for reviewing. I've committed a new patch as per your suggestions. As for the unit test, the issue was exposed in my integration tests which requires a kafka setup, any advice on how it could be done in unit tests?

**Comment by tragicjun** [comment]
Hi @suez1224, I've added some checks in the unit tests(KafkaAvroTableSourceTestBase), which would fail before the fix. Meanwhile, I found an issue in the previous version (specifically,  the produced type info should be inferred from AvroRecordClass not TableSchema), which has been fixed. Would you please take a review?

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowDeserializationSchema.java [hunk 3]
```diff
@@ -76,19 +77,35 @@
 	 */
 	private SpecificRecord record;
 
+	/** Type information describing the result type.
+	 *
+	 */
+	private final TypeInformation<Row> typeInfo;
```
Comment (twalthr) [inline flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowDeserializationSchema.java]: Declare this transient?

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowDeserializationSchema.java [hunk 4]
```diff
@@ -76,19 +77,35 @@
 	 */
 	private SpecificRecord record;
 
+	/** Type information describing the result type.
```
Comment (twalthr) [inline flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowDeserializationSchema.java]: Put the text one line down.

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowDeserializationSchema.java [hunk 5]
```diff
@@ -76,19 +77,35 @@
...
+		this(recordClazz, null);
+	}
+
+	/**
+	 * Creates a Avro deserialization schema for the given record.
+	 *
+	 * @param recordClazz Avro record class used to deserialize Avro's record to Flink's row
+	 * @param typeInfo Type information describing the result type.
+	 */
+	public AvroRowDeserializationSchema(Class<? extends SpecificRecord> recordClazz, TypeInformation<Row> typeInfo) {
```
Comment (twalthr) [inline flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroRowDeserializationSchema.java]: This constructor is not required. We should always use the result of the `AvroRecordClassConverter`.

**Comment by tragicjun** [comment]
@twalthr Thanks for the comment. I've commited a new version as per your suggestion. Particularly, I agree that TypeInfomation should always be derived from the 'avroRecrodClass', but the constructor of AvroRowDeserializationSchema only takes argument as type 'Class<? extends SpecificRecord>'. In this version I've changed the argument as 'Class<? extends SpecificRecordBase>' and fixed relevant test cases accordingly. This version of fix has passed relevant unit tests and my local integration tests.

**Comment by twalthr** [comment]
Thank you @tragicjun. LGTM. Merging...

**Comment by tragicjun** [comment]
@twalthr sure, just a second, I never did this before, let me try

**Comment by tragicjun** [comment]
@twalthr the latest commit could be merged which contains all the changes.

**Pull request closed without merging**